### PR TITLE
Refactor main compiler logic

### PR DESCRIPTION
**Pull request by iamdefinitelyahuman**
### What I did
* Refactor `vyper/compiler.py` into a subpackage
* Make the compilation process more linear
* _Make the compilation process almost 5x faster!_ :sunglasses: 

### How I did it
* Compilation is handled in five phases: ast generation, context, lll gen, assembly, bytecode
* Compiler outputs trigger these phases as data is required - nothing is repeated, nothing is generated that isn't needed

Everything is well-documented in the docstrings and module `README.md` - see the diff for more details on how it all works.

### How to verify it
Run tests.

### Cute Animal Picture
image

### CODE REVIEW

File: vyper/compiler/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+The `vyper.compiler` module contains the main user-facing functionality used to compile
+vyper source code and generate various compiler outputs.
+
+## Organization
+
+`vyper.compiler` has the following structure:
+
+* [`__init__.py`](__init__.py): Contains the `compile_codes` function, which is the
+primary function used for compiling Vyper source code.
+* [`data.py`](data.py): Pure functions for executing each compiler phase, as well as
```
Comment (fubuloubu) [inline vyper/compiler/README.md:14]: Maybe rename to `phases.py`?

File: vyper/compiler/README.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+### Compiler Phases
+
+The compilation process has five broad phases:
+
+1. In [`vyper.ast`](`../ast`), the source code is parsed and converted to an
+abstract syntax tree.
+2. The [`GlobalContext`](../parser/global_context.py) object is generated from the
+Vyper AST, analyzing and organizing the nodes prior to LLL generation.
+3. In [`vyper.parser.parser`](../parser/parser.py), the contextualized nodes are
```
Comment (fubuloubu) [inline vyper/compiler/README.md:30]: This name will change when we finish ripping apart parser, right?

File: vyper/compiler/README.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+* [`utils.py`](utils.py): Various utility functions related to compilation.
+
+## Control Flow
+
+### Compiler Phases
+
+The compilation process has five broad phases:
+
+1. In [`vyper.ast`](`../ast`), the source code is parsed and converted to an
+abstract syntax tree.
```
Comment (fubuloubu) [inline vyper/compiler/README.md:27]: I sort of want to talk about "AST nodes" briefly here (aka subclasses of `VyperNode`)

File: vyper/compiler/README.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+compiler data.
+
+## Design
+
+### Naming Conventions
+
+We use the following naming conventions throughout this module, to aid readability:
+
+* `source_code` refers to the original source code as string
+* `vyper_ast_node` refers to the top-level `Module` Vyper AST node, created from the
```
Comment (fubuloubu) [inline vyper/compiler/README.md:66]: Why not do `vyper_module` instead?

File: vyper/compiler/README.md (hunk 2 above)
Comment (iamdefinitelyahuman) [reply vyper/compiler/README.md:30]: Definitely.  It's already such a large diff, I didn't wanna start also changing things around outside.

**Pull request merged**